server/auto_updater/dataset_updater.py

```diff
@@ -0,0 +1,292 @@
+
+                    pmc_paper = open_stream.next()
+
+                    if pmc_paper is None:
+                        break
+
+                    if pmc_paper.isdir():
+                        continue
+
+                    if pmc_paper.name in current_pmc_set:
+                        continue
+
+                    new_paper = open_stream.extractfile(pmc_paper)
+                    doc_vector, word_counter = generate_vector_counts(
+                        word_model, new_paper, "//abstract/sec/*|//body/sec/*"
+                    )
+
+                    if word_counter is None:
+                        continue
+
+                    dir_writer.writerow(
+                        {"tarfile": tar_file, "file_path": str(pmc_paper.name)}
+                    )
+
+                    embed_writer.writerow(
+                        {
+                            "document": Path(pmc_paper.name).stem,
+                            "journal": str(Path(pmc_paper.name).parent),
+                            **dict(
+                                zip([f"feat_{idx}" for idx in range(300)], doc_vector)
+                            ),
+                        }
+                    )
+
+                    for tok in word_counter:
+                        count_writer.writerow(
+                            {
+                                "document": Path(pmc_paper.name).stem,
+                                "lemma": tok,
+                                "count": word_counter[tok],
+                            }
+                        )
+
+                except tarfile.ReadError as e:
+                    print(
+                        f"File {pmc_open_access_url}{tar_file} achieved an error: {str(e)}"
+                    )
+                    print("Skipping...")
+                    break
+
+            open_stream.close()
+            requested_file_stream.close()
+
+
+def generate_vector_counts(model, document_path, xpath, filter_tags=filter_tag_list):
+    """
+    This method is designed to construct document vectors for a given xml document.
+    Every document has specific tags that are striped in order to have accurate embeddings
+
+    Keyword arguments:
+        model - the model to extract word vectors from
+        xpath_str - the xpath string to extract tags from the xml document
+        filter_tag_list - the list of tags to strip from the xml document
+    """
+
+    word_vectors = []
+
+    tree = ET.parse(document_path, parser=parser)
+
+    # Process xml without specified tags
+    ET.strip_tags(tree, *filter_tags)
+
+    root = tree.getroot()
+    all_text = root.xpath(xpath)
+    all_text = list(map(lambda x: "".join(list(x.itertext())), all_text))
+    all_text = " ".join(all_text)
+
+    all_tokens = list(
+        map(
+            str,
+            filter(
+                lambda tok: str(tok) in model.wv
+                and str(tok) not in nlp.Defaults.stop_words,
+                nlp(all_text),
```

Comment:
@danich1: I measured the time of each step in this function carefully and found that this line is the bottleneck. It took about 0.6-0.8 seconds to run (depending on the size of the ppaer in XML format), which creates a `nlp` instance. After reading this page: https://spacy.io/usage/processing-pipelines#disabling
I changed line 15 from:
```
nlp = spacy.load("en_core_web_sm")
``` 
to:
```
disabled_pipelines = [
    "tagger",
    "parser",
    "ner",
    "lemmatizer",
    "attribute_ruler",
    "tok2vec",
]
nlp = spacy.load("en_core_web_sm", disable=disabled_pipelines)
```
which basically disabled all default pipelines. This change improved the processing time dramatically. Now the new papers in `comm_use.A-B.xml.tar.gz` was generated in 40 minutes (instead of more than 10 hours). I then compared the three output files: 
1. `new_papers_added_dir.tsv` is identical, this is not surprising. 
2. `new_papers_embedding.tsv` is slightly diffferent for one paper (out of 63,192 papers):
```
# "new_papers_embedding.tsv" by nlp that has no any pipelines:
BMC_Med_Genomics	PMC7011249	-0.13292737	0.29884017	-0.44909483	-0.10702868	-0.2728784	-0.27431023	0.04570517	-0.32943398	0.25982034	-0.30733883	-1.0842022	-0.31371504	0.08960449	0.048983153	-0.2921648	0.21691155	0.36423025	-0.18740693	-0.22824855	0.18023007	0.1829937	-0.17935291	0.43323973	0.5302209	0.11441889	-0.20402277	-0.71329224	-0.3292653	-0.42242202	-0.124855444	0.38212383	-0.35289082	-0.43809974	-0.040971376	0.03614347	-0.23770371	-0.73297673	-0.031328667	0.7761412	-0.6404809	0.12914778	-0.011105422	-0.13427477	-0.06248404	-0.02266382	-0.10262683	-0.054009646	-0.16534038	0.15851502	0.09253181	0.15225035	-0.19693387	-0.21560238	-0.06921824	0.13875219	0.38441402	-0.19376493	-0.14485486	0.8504384	-0.12403153	-0.34010404	-0.57511485	0.32087132	-0.4699217	-0.5991427	-0.38354787	0.07643846	0.41233927	0.4710643	0.38365448	-0.09788056	0.290843	-0.23827966	-0.08560545	-0.13456337	0.12938945	-0.12720664	0.6335732	-0.2547157	-0.13435496	0.10237407	0.015043218	0.29236546	-0.41163802	-0.47194612	0.0021375164	-0.14972048	0.81986284	0.08114276	-0.45028347	-0.17865935	0.07186784	0.3584692	-0.16232955	-0.15823059	-0.13238655	0.044880025	0.15550476	0.47688532	0.16016406	-0.48149517	-0.071176864	-0.6077426	-0.2030059	0.1768521	-0.5485124	-0.3325962	0.3548434	0.25821534	0.9249776	0.6934977	-0.011715635	-0.14146487	0.25581717	0.20474151	-0.021625146	-0.29911506	0.13406605	-0.09719923	-0.67453307	0.60133284	0.07262376	0.23028415	0.48454234	-0.09762416	0.1699019	-0.057390846	0.04708247	-0.19954087	0.23715223	0.27060398	-0.11830914	0.14309275	0.44555667	-0.59712666	0.3388609	-0.24191463	-0.14314087	-0.7870911	-0.03023561	1.0149118	-0.3531613	-0.42143822	0.7643651	-0.12799622	-0.2331133	0.35569417	0.26482323	-0.074499995	-0.33177072	-0.36383715	0.26386356	-0.5455545	-0.008056746	-0.6303604	-0.1056522	-0.04331474	0.49403423	-0.6370182	0.46350867	-0.29608604	0.767995	0.09522082	-0.6888427	-0.25943545	-0.36920825	0.2568913	0.42724705	-0.02270616	0.38679788	0.17266603	-0.062413048	0.17743604	-0.37212738	-0.42655554	0.17206773	0.32296145	0.79271144	-0.16905421	-0.076534145	0.34987396	-0.2148987	0.9390959	0.15981968	-0.14569673	0.8177112	0.09046489	-0.29979476	0.29109263	-0.25747845	0.034589037	-0.5168783	0.29158077	0.3035596	0.27107552	-0.49266487	-0.22384566	0.59415454	0.70374495	-0.03491553	0.36812606	0.1749016	0.2222262	-0.18413563	0.7522718	-0.73904186	-0.3468923	0.35301226	0.8107471	0.6857258	0.72624457	-0.04842598	0.08171677	-0.40229577	-0.19838367	0.02009173	-0.33938697	0.5002623	-0.032842625	-0.0013138345	0.00450487	-0.7297401	-0.45924833	-0.27914205	-0.5715223	-0.038493678	-0.17521471	-0.052149385	0.0069887056	0.21819152	-0.7176853	-0.16585131	0.0024012355	-0.14961186	0.24768567	0.1144673	-0.65627795	-0.30023482	0.78784555	-0.30443558	-0.43542627	-0.95847446	-0.30929092	-0.43545228	-0.4103843	-0.18132985	0.7174908	0.4473517	0.45819798	0.30763432	-0.55492395	0.17092916	-0.18448518	0.9118716	0.33605894	0.44763833	-0.38067254	-0.13120806	0.016346814	0.61980146	-0.31204784	-0.27049953	-0.17515863	0.010215435	-0.20921604	0.31492132	-0.10566855	-0.25827888	0.60265225	-0.33362845	-0.06646537	0.89532036	0.2826628	0.104968116	-0.098076604	-0.34917298	-0.114835694	0.021354351	-0.19797084	0.08269636	-0.46286005	0.25621727	-0.044358287	0.47954494	0.57841575	-0.23782842	-0.25894716	0.21093106	0.13403569	0.38822895	0.16173702	0.0679998	0.14065406	-0.2155264	-0.8014376	0.0054304083	0.2530539	0.48012	-0.264846	0.43566376

# "new_papers_embedding.tsv" by nlp that uses default pipelines:
BMC_Med_Genomics	PMC7011249	-0.13329783	0.2985041	-0.44907343	-0.10711387	-0.27270856	-0.27364478	0.04592515	-0.32981327	0.25985417	-0.3076088	-1.0842506	-0.31346402	0.08971338	0.049379833	-0.29214936	0.21728547	0.3640917	-0.18769948	-0.22824924	0.1805722	0.18349072	-0.17979616	0.43295282	0.5300225	0.11453342	-0.20423362	-0.7133477	-0.32934496	-0.4227925	-0.12463997	0.38249305	-0.3529534	-0.43800622	-0.04097121	0.0359895	-0.23720731	-0.73272455	-0.031079803	0.7759494	-0.64112175	0.12890685	-0.010842894	-0.13427377	-0.06284978	-0.022513462	-0.102368325	-0.054214925	-0.16526884	0.1583605	0.09270608	0.1523684	-0.19713528	-0.21520148	-0.06848164	0.13854071	0.38490188	-0.19355884	-0.14503074	0.8504772	-0.12428724	-0.33967698	-0.575194	0.3211894	-0.46972448	-0.5989744	-0.38323897	0.076483205	0.41214877	0.47124505	0.38379228	-0.0981191	0.2906242	-0.2384808	-0.08514096	-0.13465413	0.12927994	-0.12736182	0.63331974	-0.2546803	-0.13434272	0.10246605	0.015231185	0.2921538	-0.41120636	-0.4712843	0.0017976207	-0.14948858	0.81935096	0.08085463	-0.45072702	-0.17916454	0.07177641	0.3586952	-0.16248855	-0.15809666	-0.13172768	0.04468074	0.15562668	0.47649956	0.16003174	-0.4813163	-0.071412496	-0.6074932	-0.202922	0.17691574	-0.5481237	-0.33243045	0.35418504	0.25869006	0.9247417	0.69340324	-0.011140373	-0.14154555	0.2558156	0.20448682	-0.02119166	-0.29957274	0.13454105	-0.09715195	-0.6750654	0.60124123	0.071792744	0.2301529	0.4845385	-0.09749041	0.17005405	-0.057398222	0.047417544	-0.19960637	0.23718034	0.27057785	-0.118096456	0.14269422	0.44519046	-0.59746414	0.33925036	-0.24228315	-0.14289501	-0.78716975	-0.030330522	1.0147663	-0.35318717	-0.42100698	0.7648233	-0.12801486	-0.23308009	0.35571018	0.26512355	-0.07438472	-0.33178163	-0.36325145	0.26371768	-0.5456856	-0.0080338335	-0.63013744	-0.105526745	-0.043356992	0.49423888	-0.6372714	0.4636167	-0.2962501	0.7680732	0.095363386	-0.68890685	-0.2592665	-0.36923698	0.25673106	0.42716873	-0.02258582	0.38716576	0.172638	-0.06250487	0.17791887	-0.3716145	-0.42696643	0.17202722	0.32335797	0.7920022	-0.16948284	-0.07673464	0.3505372	-0.21438234	0.93847114	0.15991983	-0.14584404	0.81803983	0.09041629	-0.30031037	0.2905811	-0.257404	0.033958886	-0.5169393	0.2918442	0.30326852	0.2708712	-0.49251148	-0.22351687	0.5946208	0.70332026	-0.03472063	0.36786166	0.17489213	0.22210799	-0.18444033	0.7525716	-0.7393359	-0.3471673	0.352535	0.810382	0.68553257	0.7260217	-0.04827968	0.08188952	-0.40206853	-0.19833997	0.020386191	-0.33923265	0.5002762	-0.03291202	-0.0012371549	0.004156608	-0.7295617	-0.45917565	-0.27913734	-0.5713705	-0.038378604	-0.17486218	-0.051961567	0.0074344287	0.21759564	-0.71781355	-0.1655077	0.0032356249	-0.14924853	0.24773426	0.11482885	-0.6564267	-0.30005544	0.78735924	-0.30432242	-0.43576467	-0.95888114	-0.308755	-0.43565586	-0.4103519	-0.18126269	0.71747774	0.4477098	0.45848978	0.30780977	-0.5548276	0.1705107	-0.18479058	0.9121545	0.33627686	0.44746223	-0.38045883	-0.1308282	0.016342884	0.6194594	-0.3123457	-0.27110294	-0.17504577	0.010118301	-0.20896217	0.31456918	-0.10520819	-0.25849873	0.6024002	-0.33340755	-0.06645426	0.89524645	0.28265885	0.103977345	-0.097458385	-0.34895888	-0.115495145	0.021388829	-0.19837083	0.08247762	-0.46293753	0.2564294	-0.04381285	0.47947103	0.5779067	-0.23786744	-0.2597661	0.2114751	0.13404949	0.38776726	0.16205749	0.06802507	0.1404274	-0.21495517	-0.80140054	0.0055945283	0.25279152	0.48017228	-0.26451328	0.4358428
```
3. `new_paper_token_count.tsv` is slightly different too for the same paper:
```
# "new_paper_token_count.tsv" by nlp that has no pipelines:
# ----------------------------------------------------------
PMC7011249	.	395
PMC7011249	p.	68
PMC7011249	p	21
# ...

# "new_paper_token_count.tsv" by nlp that uses default pipelines:
# -----------------------------------------------------------------
PMC7011249	.	394
PMC7011249	p.	69
PMC7011249	p	20
# ...
```
The differences must be due to the disabled pipelines. I am not an expert on NLP in spaCy. What do you think?

Reply:
I'm all for disabling non-necessary features. Good find!